kroxylicious/src/main/java/io/kroxylicious/proxy/internal/net/EndpointRegistry.java

```diff
@@ -0,0 +1,417 @@
+                        .handle((unused1, t) -> {
+                            registeredVirtualClusters.remove(virtualCluster);
+                            if (t != null) {
+                                deregisterFuture.completeExceptionally(t);
+                            }
+                            else {
+                                deregisterFuture.complete(null);
+                            }
+                            return null;
+                        }));
+        return deregisterFuture;
+    }
+
+    /* test */ boolean isRegistered(VirtualCluster virtualCluster) {
+        return registeredVirtualClusters.containsKey(virtualCluster);
+    }
+
+    /* test */ int listeningChannelCount() {
+        return listeningChannels.size();
+    }
+
+    private CompletionStage<Endpoint> registerBinding(Endpoint key, String host, VirtualClusterBinding virtualClusterBinding) {
+        Objects.requireNonNull(key, "key cannot be null");
+        Objects.requireNonNull(virtualClusterBinding, "virtualClusterBinding cannot be null");
+        var virtualCluster = virtualClusterBinding.virtualCluster();
+
+        var lcr = listeningChannels.computeIfAbsent(key, k -> {
+            // the listening channel doesn't exist, atomically create a record to represent it and request its binding to the network.
+            var future = new CompletableFuture<Channel>();
+            var r = ListeningChannelRecord.create(future.exceptionally(t -> {
+                // Handles the case where the network bind fails
+                listeningChannels.remove(key);
+                if (t instanceof RuntimeException re) {
+                    throw re;
+                }
+                else {
+                    throw new RuntimeException(t);
+                }
+            }));
+
+            queue.add(new NetworkBindRequest(key.bindingAddress(), key.port(), virtualCluster.isUseTls(), future));
+            return r;
+        });
+
+        if (lcr.unbindingStage().get() != null) {
+            // Listening channel already being unbound, chain this request to the unbind stage so it completes later.
+            return lcr.unbindingStage().get().thenCompose(u -> registerBinding(key, host, virtualClusterBinding));
+        }
+
+        return lcr.bindingStage().thenApply(c -> {
+            synchronized (lcr) {
+                var bindings = c.attr(CHANNEL_BINDINGS);
+                bindings.setIfAbsent(new ConcurrentHashMap<>());
+                var bindingMap = bindings.get();
+                var bindingKey = virtualCluster.requiresTls() ? RoutingKey.createBindingKey(host) : RoutingKey.NULL_ROUTING_KEY;
+
+                var existing = bindingMap.putIfAbsent(bindingKey, virtualClusterBinding);
+
+                if (existing != null) {
+                    throw new EndpointBindingException("Endpoint %s cannot be bound with key %s, that key is already bound".formatted(key, bindingKey));
```

Comment:
Would it be worth looking up and logging the clashing virtual clusters?

Reply:
I've logged the binding. That's not ideal was the virtual cluster currently doesn't know its own name.  I don't want to fix that as part of this PR.

Reply:
Yeah that seems like something we should fix separately